adds ability to repeat over object keys

pflannery opened the pull request:
relates with #116
relates with https://github.com/Polymer/polymer-expressions/issues/11

example syntax:

``` html
<template id="KeysTemplate" is="auto-binding">
  <template repeat="{{key in objectData}}">
    <div>{{key}} - {{objectData[key]}}</div>
  </template>
</template>
```

pflannery committed:
adds ability to repeat over object keys

adds the following tests:

Repeat If (Object Keys)
Repeat If (false-true-false)
Repeat If (false-true-false) (Object Keys)
Repeat oneTime-If (predicate false) (Object Keys)
Repeat oneTime-If (predicate true) (Object Keys)
oneTime-Repeat If (Object Keys)
oneTime-Repeat oneTime-If (Object Keys)
Repeat (Object Keys)
Repeat - oneTime (Object Keys)

pflannery commented:
also relates with https://github.com/Polymer/polymer/issues/1062

pflannery commented:
@rafaelw PTAL

jmesserly commented on src/TemplateBinding.js:
```diff
@@ -99,6 +99,53 @@
+  // Use for template repeating objects
+  function ObjectArrayProxy(values, observe, onAddedCallback, onRemovedCallback) {
+    this.values = values;
+    this.onAdded = onAddedCallback;
```
is this callback used?

pflannery replied:
no, I will remove.

jmesserly commented on src/TemplateBinding.js:
```diff
@@ -1058,6 +1106,20 @@
     updateValue: function(value) {
       if (!this.deps.repeat)
         value = [value];
+      else if (Object.prototype.toString.call(value)  === '[object Object]') {
```
what is the intent of this test?

pflannery replied:
this ensures that it's an object

jmesserly replied:
right, but what does "an object" mean? :)

It doesn't seem great to rely on Object.prototype.toString here ... I'm wondering if there's a more conventional way to express the check. e.g.

```js
if (value != null && typeof value == 'object') { ... }
```

pflannery replied:
if the value is an Array then ```(typeof value == 'object') === true```

jmesserly commented on src/TemplateBinding.js:
```diff
@@ -1058,6 +1106,20 @@
       if (!this.deps.repeat)
         value = [value];
+      else if (Object.prototype.toString.call(value)  === '[object Object]') {
+        if (this.objectArrayProxy !== undefined) {
```
`if (!this.objectArrayProxy)` to match style used elsewhere

jmesserly commented on src/TemplateBinding.js:
```diff
@@ -1058,6 +1106,20 @@
+      else if (Object.prototype.toString.call(value)  === '[object Object]') {
+        if (this.objectArrayProxy !== undefined) {
+          if (value !== this.objectArrayProxy.values) {
+            this.objectArrayProxy.closeObserver();
```
it seems like some duplication between closeObserver and resetKeys_ (which itself knows how to do closeObserver), is there a way to factor that better? ideally the code here is more like:

```js
if (!this.keysObserver) {
  this.keysObserver = new ...
}
// start or reset observation on this.keysObserver
```

jmesserly commented on src/TemplateBinding.js:
```diff
@@ -99,6 +99,53 @@
   }
 
+  // Use for template repeating objects
+  function ObjectArrayProxy(values, observe, onAddedCallback, onRemovedCallback) {
```
Maybe ObjectKeysObserver? ("proxy" means something specific in ES6 ... "ArrayProxy" made me think it was a proxy for an array ... whereas the object's keys is established terminology, e.g. [Object.keys](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys))

jmesserly commented on src/TemplateBinding.js:
```diff
@@ -1058,6 +1106,20 @@
         value = [value];
+      else if (Object.prototype.toString.call(value)  === '[object Object]') {
+        if (this.objectArrayProxy !== undefined) {
+          if (value !== this.objectArrayProxy.values) {
```
it would be nice to have a test to validate that this check can succeed, in other words, that we aren't allocating a new observer every time.

jmesserly commented on test/tests.js:
```diff
@@ -1072,6 +1072,53 @@ suite('Template Instantiation', function() {
+    var div = createTestHtml(
+      '<template repeat="{{ obj }}"">{{}}</template>');
+
+    var model = { obj: {"key1": 1, "key2": 2, "key3": 3} };
```
thoughts on supporting ES6 Map? http://kangax.github.io/compat-table/es6/#Map
it's going to replace objects as the main way of doing maps (at least, that is what the EcmaScript designers are hoping :) )

pflannery replied:
see my ES6 map thoughts below

jmesserly commented:
I added some comments. Overall it looks like it would work. One high level comment, did you consider a design that adds support to https://github.com/Polymer/polymer-expressions instead of TemplateBinding? I think the desire was to fix it at that layer (hence moving this issue https://github.com/Polymer/polymer-expressions/issues/11 to polymer-expressions)

pflannery commented:
Best to change it in TemplateBinding because that is where we can determine if we have to observe and repeat an Object or an Array at runtime.

I did want to make a change at PolymerExpressions as well so we can also use expressions like ```{{key, value of obj}}``` but to do this we need to amend Esprima and PolymerExpressions with two simple methods called parseOfExpression and createOfExpression.
An alternative would be to add an extra attribute like ```<template repeat='{{key, value in object}}' repeat-values>``` but that isn't as sexy.

Using an ES6 map in this case wouldn't be possible as its implementation uses iterators to obtain the keys and values meaning we cant monitor keys or values being added or removed. If only there was an ES6 map observer or change callback.

pflannery commented:
@jmesserly squashed commits back to one. Added 7 more unit tests (cloned from existing tests and transformed to handle object keys)
I found and fixed an issue where keys weren't being applied to the value during an if operation so added a unit test case 'Repeat If (false-true-false) (Object Keys)'

PTAL

pflannery committed:
renames repeat if predicate false-true-false tests

pflannery commented:
closed in favour of implementation in 0.8 https://github.com/Polymer/polymer/issues/1385
